File: 1036-rotting-oranges/rotting-oranges.cpp
class Solution {
    
public:
    int orangesRotting(vector<vector<int>>& grid) {


        int fresh=0;
        int time=0;
        int rot=0;
vector<pair<int, int>> directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        queue<pair<int,int>>q;
        for(int r=0;r<grid.size();r++)
        {
            for(int c=0;c<grid[0].size();c++)
            {
                if(grid[r][c]==1)
                {
                    fresh++;
                }
                else if(grid[r][c]==2)
                {
                    
                    q.push({r,c});
                }
            }
            
        }


        while(fresh>0 && !q.empty())
        {
            int size=q.size();
            for(int i=0;i<size;i++)
            {
                auto[row,col]=q.front();
                q.pop();

            for(const auto&dir: directions)
            {
                int nr=row+dir.first;
                int nc=col+dir.second;

                 
                    

                    if(nr>=0 && nc>=0 && nr<grid.size() && nc<grid[0].size() && grid[nr][nc]==1)
                    {
                        grid[nr][nc]=2;
                        q.push({nr,nc});
                        fresh--;

                    }
                


            }
            
            }
            time++;
            
        }
return fresh == 0 ? time : -1;
        
        
    }
};

File: 1036-rotting-oranges/rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        #basically first find all the rotten oranges and start rotting from there

        row=len(grid)
        col=len(grid[0])
        q=deque()
        fresh=0
        rotten=0
        directions=[[1,0],[-1,0],[0,1],[0,-1]]
        for r in range(row):
            for c in range(col):
                if grid[r][c]==2:
                    q.append([r,c])
                    rotten+=1
                if grid[r][c]==1:
                    fresh+=1
        time=0
        while fresh!=0 and q:
            l=len(q)
            for i in range(l):
                node=q.popleft()
                for r,c in directions:
                    nr=node[0]+r
                    nc=node[1]+c
                    if nr in range(len(grid)) and nc in range(len(grid[0])) and grid[nr][nc]==1:
                        grid[nr][nc]=2
                        q.append([nr,nc])
                        rotten+=1
                        fresh-=1
            time+=1
        return time if fresh==0 else -1


        

        